### ISSUE TITLE
`chain-wallet-libs` private payload creation failing with `Cannot parse public key bytes`

### ISSUE THREAD

**Issue by refi93** [input-output-hk/catalyst-core#75]
I'm trying to create a private vote payload and it's failing with a `Cannot parse public key bytes` error, here's the snippet:

```javascript
const lib = await import('wallet-js')

const votePlanId = lib.VotePlanId.from_bytes(
  Buffer.from(
    'deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef',
    'hex',
  ),
)
const optionsCount = 2
const choice = 47
const pubKey = Buffer.from(
  '62def90b803d2c5616165ac40db911b98261bdc2c2a9678a0a3837024fa2f998',
  'hex'
)

// fails with "Cannot parse public key bytes" error
lib.Payload.new_private(votePlanId, optionsCount, choice, pubKey)
```

creation of a public payload seems to work fine and I don't see the snippet above differing in any significant way from the unit test in the lib repo, at least the lengths seem to be right and the public key above was derived as a proper ed25519 public key (non-extended): [https://github.com/input-output-hk/catalyst-core/blob/28690a38d9ab34802bd4afed173d[…]/chain-wallet-libs/bindings/wallet-js/js-test/test/vote_cast.js](https://github.com/input-output-hk/catalyst-core/blob/28690a38d9ab34802bd4afed173d39015156fb84/src/chain-wallet-libs/bindings/wallet-js/js-test/test/vote_cast.js#L111)

Could you advise? Also, my assumption is that the `pubKey` is the user's voting public key - is that correct?

note: The public key from the unit test (`bed88887abe0a84f64691fe0bdfa3daf1a6cd697a13f07ae07588910ce39c927`) fails with `Unreachable` error

**Comment by minikin** [comment]
@dkijania can you have a look, please?

**Comment by dkijania** [comment]
Hello @refi93 ,

Indeed `new_private` method consumes public key bytes as last argument. However, it is not wallet public key but election public key (which will be used for vote encryption). I decoded your key and got:

```
echo 62def90b803d2c5616165ac40db911b98261bdc2c2a9678a0a3837024fa2f998 | jcli key from-bytes --public --type ed25519
ed25519_pk1vt00jzuq85k9v9skttzqmwg3hxpxr0wzc25k0zs28qmsynazlxvqv4xljp
```

Which confirms that it is ed25519 used for signing transaction. However, we need to encrypt also vote so only committees can decrypt it with election key (which is combined signature of committee members).

Election key looks like this:

 `ristretto255_votepk1yzvwlnk80yqs2kmve5uj3p0c52cpghmep940vad5zlnluzpxfqeqfj7fg4`

Once you have it you need to convert it to bytes and pass it to method, similar to below snippet (apologies for RUST code):

```
let key = ElectionPublicKey::try_from_bech32_str("ristretto255_votepk1yzvwlnk80yqs2kmve5uj3p0c52cpghmep940vad5zlnluzpxfqeqfj7fg4").unwrap();
let bytes = key.to_bytes();
```

Now, there are two questions: how to get election key?  and how to use it in JS?

Election public key is available from catalyst backend api and is different for each voting event. This can be read from REST API: 

`https://{CATALYST_BACKEND_API}/api/v0/fund` 

and then to get particular voteplan id  you can find it by xpath:

`$.chain_vote_plans[?(@.chain_voteplan_id = 'xxxx')].chain_vote_encryption_key`

Currently ElectionPublicKey is not exposed to JS so cannot be easily used. Maybe there is some JS library to consume custom HRP and convert to bytes. Maybe @Mr-Leshiy  knows ?

**Comment by Mr-Leshiy** [comment]
Hello @refi93 !

@dkijania is correct, you should sign private payload using another type of key. Unfortunately don't know about any such JS library, but if needed I can provide any needed and convenient functionality for ElectionPublicKey that you need @refi93 . What do you think what can be missed there ? 

> note: The public key from the unit test (bed88887abe0a84f64691fe0bdfa3daf1a6cd697a13f07ae07588910ce39c927) fails with Unreachable error

The problem is that `choice` value should be less than `optionsCount`. I have updated error messaging so it will be more informative and clear.
```
const optionsCount = 2
const choice = 47
assert(choice < optionsCount)
```

**Comment by Scitz0** [comment]
If election public key is available from catalyst backend api, I think this should be provided in the cip-62 signVotes call dApp make to wallet. 

And we would need to find a js solution like you mentioned.

**Comment by minikin** [comment]
The issue was addressed in #125.
